src/core/cycle_detector.go

```diff
@@ -20,26 +20,35 @@ func (c dependencyChain) String() string {
 }
 
 type cycleDetector struct {
-	deps  map[*BuildLabel][]*BuildLabel
-	queue chan dependencyLink
+	deps        map[*BuildLabel]map[*BuildLabel]struct{}
```

Comment:
Might be worth noting what exactly this map represents (like what are the keys/values in each one)